Retry on retrieving team slug after creation

<!--
Clear description of success criteria. Verbosity is ✨!
-->
@mickeygousset was testing out a migration and `create-team` command failed. It appears it called an API to create a new team, and got back a success. Then tried to call an API to get the team details, and GitHub said no teams exist.

This feels like some kind of race condition in the GH API, and I suspect if we waited a few seconds then tried to retrieve the team again it would work.

Our existing retry logic didn't kick in here, because the call to get the team details didn't fail, it succeeded it just didn't include the data in the response (the new team) that we expected it to.

```
[11:34 AM] [DEBUG] HTTP GET: https://raw.githubusercontent.com/github/gh-gei/main/LATEST-VERSION.txt
[11:34 AM] [DEBUG] RESPONSE (OK): v0.26

[11:34 AM] [INFO] You are running the latest version of the ado2gh CLI [v0.26]
[11:34 AM] [INFO] Creating GitHub team...
[11:34 AM] [INFO] GITHUB ORG: mickey-test-migrations
[11:34 AM] [INFO] TEAM NAME: MyShuttle-Maintainers
[11:34 AM] [INFO] IDP GROUP: MyShuttle-Maintainers
[11:34 AM] [DEBUG] HTTP GET: https://api.github.com/orgs/mickey-test-migrations/teams
[11:34 AM] [DEBUG] GITHUB REQUEST ID: E454:659E:BD4178:182F6A6:6329EB9E
[11:34 AM] [DEBUG] RESPONSE (OK): []
[11:34 AM] [DEBUG] HTTP POST: https://api.github.com/orgs/mickey-test-migrations/teams
[11:34 AM] [DEBUG] HTTP BODY: {"name":"MyShuttle-Maintainers","privacy":"closed"}
[11:34 AM] [DEBUG] GITHUB REQUEST ID: E454:659E:BD41D8:182F76A:6329EB9F
[11:34 AM] [DEBUG] RESPONSE (Created): {"name":"MyShuttle-Maintainers","id":6656283,"node_id":"T_kwDOBsS6Ms4AZZEb","slug":"myshuttle-maintainers","description":null,"privacy":"closed","url":"https://api.github.com/organizations/113556018/team/6656283","html_url":"https://github.com/orgs/mickey-test-migrations/teams/myshuttle-maintainers","members_url":"https://api.github.com/organizations/113556018/team/6656283/members{/member}","repositories_url":"https://api.github.com/organizations/113556018/team/6656283/repos","permission":"pull","parent":null,"created_at":"2022-09-20T16:34:39Z","updated_at":"2022-09-20T16:34:39Z","members_count":1,"repos_count":0,"organization":{"login":"mickey-test-migrations","id":113556018,"node_id":"O_kgDOBsS6Mg","url":"https://api.github.com/orgs/mickey-test-migrations","repos_url":"https://api.github.com/orgs/mickey-test-migrations/repos","events_url":"https://api.github.com/orgs/mickey-test-migrations/events","hooks_url":"https://api.github.com/orgs/mickey-test-migrations/hooks","issues_url":"https://api.github.com/orgs/mickey-test-migrations/issues","members_url":"https://api.github.com/orgs/mickey-test-migrations/members{/member}","public_members_url":"https://api.github.com/orgs/mickey-test-migrations/public_members{/member}","avatar_url":"https://avatars.githubusercontent.com/u/113556018?v=4","description":null,"is_verified":false,"has_organization_projects":true,"has_repository_projects":true,"public_repos":0,"public_gists":0,"followers":0,"following":0,"html_url":"https://github.com/mickey-test-migrations","created_at":"2022-09-14T20:35:08Z","updated_at":"2022-09-19T20:05:37Z","type":"Organization"}}
[11:34 AM] [INFO] Successfully created team
[11:34 AM] [DEBUG] HTTP GET: https://api.github.com/orgs/mickey-test-migrations/teams
[11:34 AM] [DEBUG] GITHUB REQUEST ID: E454:659E:BD42F1:182F9A4:6329EBA0
[11:34 AM] [DEBUG] RESPONSE (OK): []
[11:34 AM] [ERROR] System.InvalidOperationException: Source sequence doesn't contain any elements.
   at System.Linq.AsyncEnumerable.<SingleAsync>g__Core|335_0[TSource](IAsyncEnumerable`1 source, Func`2 predicate, CancellationToken cancellationToken) in /_/Ix.NET/Source/System.Linq.Async/System/Linq/Operators/Single.cs:line 101
   at OctoshiftCLI.GithubApi.GetTeamSlug(String org, String teamName)
   at OctoshiftCLI.Commands.CreateTeamCommandBase.Handle(String githubOrg, String teamName, String idpGroup, String githubPat, Boolean verbose)
   at System.CommandLine.Invocation.CommandHandler.GetExitCodeAsync(Object value, InvocationContext context)
   at System.CommandLine.Invocation.ModelBindingCommandHandler.InvokeAsync(InvocationContext context)
   at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass23_0.<<UseParseErrorReporting>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<<UseHelp>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass27_0.<<UseVersionOption>b__1>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass25_0.<<UseTypoCorrections>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__24_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass22_0.<<UseParseDirective>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass11_0.<<UseDebugDirective>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<RegisterWithDotnetSuggest>b__10_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass14_0.<<UseExceptionHandler>b__0>d.MoveNext()
```

--- a/RELEASENOTES.md
+++ b/RELEASENOTES.md
@@ -1,2 +1,3 @@
 - Update `gh bbs2gh generate-script` so it supports more than 25 projects/repos
 - Rename `--bbs-project-key` to `--bbs-project` in `gh bbs2gh generate-script` for consistency
+- Fix a bug where `create-team` might not work due to a race condition

--- a/src/Octoshift/GithubApi.cs
+++ b/src/Octoshift/GithubApi.cs
@@ -65,7 +65,7 @@ namespace OctoshiftCLI
             await _client.DeleteAsync(url);
         }
 
-        public virtual async Task<string> CreateTeam(string org, string teamName)
+        public virtual async Task<(string Id, string Slug)> CreateTeam(string org, string teamName)
         {
             var url = $"{_apiUrl}/orgs/{org}/teams";
             var payload = new { name = teamName, privacy = "closed" };
@@ -73,15 +73,15 @@ namespace OctoshiftCLI
             var response = await _client.PostAsync(url, payload);
             var data = JObject.Parse(response);
 
-            return (string)data["id"];
+            return ((string)data["id"], (string)data["slug"]);
         }
 
-        public virtual async Task<IEnumerable<string>> GetTeams(string org)
+        public virtual async Task<IEnumerable<(string Name, string Slug)>> GetTeams(string org)
         {
             var url = $"{_apiUrl}/orgs/{org}/teams";
 
             return await _client.GetAllAsync(url)
-                .Select(t => (string)t["name"])
+                .Select(t => ((string)t["name"], (string)t["slug"]))
                 .ToListAsync();
         }
 

--- a/src/Octoshift/Handlers/CreateTeamCommandHandler.cs
+++ b/src/Octoshift/Handlers/CreateTeamCommandHandler.cs
@@ -2,6 +2,7 @@ using System;
 using System.Linq;
 using System.Threading.Tasks;
 using OctoshiftCLI.Commands;
+using OctoshiftCLI.Extensions;
 
 namespace OctoshiftCLI.Handlers;
 
@@ -38,18 +39,16 @@ public class CreateTeamCommandHandler : ICommandHandler<CreateTeamCommandArgs>
         _log.RegisterSecret(args.GithubPat);
 
         var teams = await _githubApi.GetTeams(args.GithubOrg);
-        if (teams.Contains(args.TeamName))
+        var teamSlug = teams.FirstOrDefault(t => t.Name == args.TeamName).Slug;
+        if (teamSlug.HasValue())
         {
-            _log.LogSuccess($"Team '{args.TeamName}' already exists - New team will not be created");
+            _log.LogSuccess($"Team '{args.TeamName}' already exists. New team will not be created");
         }
         else
         {
-            await _githubApi.CreateTeam(args.GithubOrg, args.TeamName);
+            (_, teamSlug) = await _githubApi.CreateTeam(args.GithubOrg, args.TeamName);
             _log.LogSuccess("Successfully created team");
         }
-
-        // TODO: Can improve perf by capturing slug in the response from CreateTeam or GetTeams
-        var teamSlug = await _githubApi.GetTeamSlug(args.GithubOrg, args.TeamName);
 
         if (string.IsNullOrWhiteSpace(args.IdpGroup))
         {

--- a/src/OctoshiftCLI.Tests/GithubApiTests.cs
+++ b/src/OctoshiftCLI.Tests/GithubApiTests.cs
@@ -128,7 +128,7 @@ namespace OctoshiftCLI.Tests
         }
 
         [Fact]
-        public async Task CreateTeam_Returns_Created_Team_Id()
+        public async Task CreateTeam_Returns_Created_Team_Id_And_Slug()
         {
             // Arrange
             const string teamName = "TEAM_NAME";
@@ -137,7 +137,8 @@ namespace OctoshiftCLI.Tests
             var payload = new { name = teamName, privacy = "closed" };
 
             const string teamId = "TEAM_ID";
-            var response = $"{{\"id\": \"{teamId}\"}}";
+            const string teamSlug = "TEAM_SLUG";
+            var response = $"{{\"id\": \"{teamId}\", \"slug\": \"{teamSlug}\"}}";
 
             _githubClientMock
                 .Setup(m => m.PostAsync(url, It.Is<object>(x => x.ToJson() == payload.ToJson()), null))
@@ -147,7 +148,7 @@ namespace OctoshiftCLI.Tests
             var result = await _githubApi.CreateTeam(GITHUB_ORG, teamName);
 
             // Assert
-            result.Should().Be(teamId);
+            result.Should().Be((teamId, teamSlug));
         }
 
         [Fact]
@@ -156,17 +157,17 @@ namespace OctoshiftCLI.Tests
             // Arrange
             var url = $"https://api.github.com/orgs/{GITHUB_ORG}/teams";
 
-            const string team1 = "TEAM_1";
-            const string team2 = "TEAM_2";
-            const string team3 = "TEAM_3";
-            const string team4 = "TEAM_4";
+            var team1 = (Name: "TEAM_1", Slug: "SLUG_1");
+            var team2 = (Name: "TEAM_2", Slug: "SLUG_2");
+            var team3 = (Name: "TEAM_3", Slug: "SLUG_3");
+            var team4 = (Name: "TEAM_4", Slug: "SLUG_4");
 
             var teamsResult = new[]
             {
-                new { id = 1, name = team1 },
-                new { id = 2, name = team2 },
-                new { id = 3, name = team3 },
-                new { id = 4, name = team4 }
+                new { id = 1, name = team1.Name, slug = team1.Slug },
+                new { id = 2, name = team2.Name, slug = team2.Slug },
+                new { id = 3, name = team3.Name, slug = team3.Slug },
+                new { id = 4, name = team4.Name, slug = team4.Slug }
             }.ToAsyncJTokenEnumerable();
 
             _githubClientMock

--- a/src/OctoshiftCLI.Tests/Octoshift/Handlers/CreateTeamCommandHandlerTests.cs
+++ b/src/OctoshiftCLI.Tests/Octoshift/Handlers/CreateTeamCommandHandlerTests.cs
@@ -1,5 +1,6 @@
 using System.Collections.Generic;
 using System.Threading.Tasks;
+using FluentAssertions;
 using Moq;
 using OctoshiftCLI.Commands;
 using OctoshiftCLI.Handlers;
@@ -31,8 +32,8 @@ public class CreateTeamCommandHandlerTests
     {
         _mockGithubApi.Setup(x => x.GetTeamMembers(GITHUB_ORG, TEAM_SLUG).Result).Returns(TEAM_MEMBERS);
         _mockGithubApi.Setup(x => x.GetIdpGroupId(GITHUB_ORG, IDP_GROUP).Result).Returns(IDP_GROUP_ID);
-        _mockGithubApi.Setup(x => x.GetTeamSlug(GITHUB_ORG, TEAM_NAME).Result).Returns(TEAM_SLUG);
-        _mockGithubApi.Setup(x => x.GetTeams(GITHUB_ORG).Result).Returns(new List<string>());
+        _mockGithubApi.Setup(x => x.CreateTeam(GITHUB_ORG, TEAM_NAME).Result).Returns(("1", TEAM_SLUG));
+        _mockGithubApi.Setup(x => x.GetTeams(GITHUB_ORG).Result).Returns(new List<(string, string)>());
 
         var args = new CreateTeamCommandArgs
         {
@@ -53,8 +54,7 @@ public class CreateTeamCommandHandlerTests
     {
         _mockGithubApi.Setup(x => x.GetTeamMembers(GITHUB_ORG, TEAM_SLUG).Result).Returns(TEAM_MEMBERS);
         _mockGithubApi.Setup(x => x.GetIdpGroupId(GITHUB_ORG, IDP_GROUP).Result).Returns(IDP_GROUP_ID);
-        _mockGithubApi.Setup(x => x.GetTeamSlug(GITHUB_ORG, TEAM_NAME).Result).Returns(TEAM_SLUG);
-        _mockGithubApi.Setup(x => x.GetTeams(GITHUB_ORG).Result).Returns(new List<string> { TEAM_NAME });
+        _mockGithubApi.Setup(x => x.GetTeams(GITHUB_ORG).Result).Returns(new List<(string, string)> { (TEAM_NAME, TEAM_SLUG) });
 
         var actualLogOutput = new List<string>();
         _mockOctoLogger.Setup(m => m.LogInformation(It.IsAny<string>())).Callback<string>(s => actualLogOutput.Add(s));
@@ -72,6 +72,6 @@ public class CreateTeamCommandHandlerTests
         _mockGithubApi.Verify(x => x.RemoveTeamMember(GITHUB_ORG, TEAM_SLUG, TEAM_MEMBERS[0]));
         _mockGithubApi.Verify(x => x.RemoveTeamMember(GITHUB_ORG, TEAM_SLUG, TEAM_MEMBERS[1]));
         _mockGithubApi.Verify(x => x.AddEmuGroupToTeam(GITHUB_ORG, TEAM_SLUG, IDP_GROUP_ID));
-        actualLogOutput.Contains($"Team '{TEAM_NAME}' already exists. New team will not be created");
+        actualLogOutput.Should().Contain($"Team '{TEAM_NAME}' already exists. New team will not be created");
     }
 }

